File: src/algorithm/generic/solution121/Solution.java
package algorithm.generic.solution121;

/**
 * https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
 */
public class Solution {
    public int maxProfit(int[] prices) {
        int min = Integer.MAX_VALUE;
        int maxProfit = 0;
        
        for(int i=0; i< prices.length; i++) {
        	maxProfit = Math.max(maxProfit, prices[i] - min);
        	min = Math.min(min, prices[i]);
        	
        }
        
        return maxProfit;
    }
    
	public static void main(String[] args) {
		Solution solution = new Solution();
		System.out.println(solution.maxProfit(new int[] {7,1,5,3,6,4}));
	}

}

File: bin/algorithm/generic/solution121/Solution.py
from typing import List
import sys;


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        min_price = sys.maxsize
        max_profit = 0
        for price in prices:
            max_profit = max(max_profit, price - min_price)
            min_price = min(min_price, price)
        
        return max_profit

def main():
    solution = Solution()
    print(solution.maxProfit([7,1,5,3,6,4]))

if __name__ == '__main__':
    main()